fluss-server/src/main/java/org/apache/fluss/server/kv/rowmerger/AggregateRowMerger.java

```diff
@@ -0,0 +1,399 @@
+ *
+ * <p>This merger supports schema evolution by dynamically retrieving schemas based on schema IDs
+ * when merging rows with different schema versions.
+ *
+ * <p>This class is thread-safe as it is guaranteed to be accessed by a single thread at a time
+ * (protected by KvTablet's write lock).
+ */
+public class AggregateRowMerger implements RowMerger {
+
+    // Cache configuration constants
+    private static final int PARTIAL_MERGER_CACHE_MAX_SIZE = 4;
+    private static final Duration PARTIAL_MERGER_CACHE_EXPIRE_DURATION = Duration.ofMinutes(5);
+
+    private final SchemaGetter schemaGetter;
+    private final DeleteBehavior deleteBehavior;
+    private final AggregationContextCache contextCache;
+
+    // Cache for PartialAggregateRowMerger instances to avoid repeated creation
+    private final Cache<CacheKey, PartialAggregateRowMerger> partialMergerCache;
+
+    public AggregateRowMerger(
+            TableConfig tableConfig, KvFormat kvFormat, SchemaGetter schemaGetter) {
+        this.schemaGetter = schemaGetter;
+        // Extract configuration from TableConfig to ensure single source of truth
+        this.deleteBehavior = tableConfig.getDeleteBehavior().orElse(DeleteBehavior.IGNORE);
+        this.contextCache = new AggregationContextCache(schemaGetter, kvFormat);
+        // Initialize cache with same settings as PartialUpdaterCache and AggregationContextCache
+        this.partialMergerCache =
+                Caffeine.newBuilder()
+                        .maximumSize(PARTIAL_MERGER_CACHE_MAX_SIZE)
+                        .expireAfterAccess(PARTIAL_MERGER_CACHE_EXPIRE_DURATION)
+                        .build();
+    }
+
+    @Override
+    public BinaryValue merge(BinaryValue oldValue, BinaryValue newValue) {
+        // First write: no existing row
+        if (oldValue == null || oldValue.row == null) {
+            return newValue;
+        }
+
+        // Get contexts for schema evolution support
+        // Optimize: reuse context when schemaId is the same, and avoid Tuple2 allocation
+        AggregationContext oldContext;
+        AggregationContext newContext;
+
+        if (oldValue.schemaId == newValue.schemaId) {
+            // Same schema: reuse the same context to avoid duplicate lookup
+            newContext = contextCache.getContext(newValue.schemaId);
+            oldContext = newContext;
+        } else {
+            // Different schemas: need separate lookups
+            oldContext = contextCache.getContext(oldValue.schemaId);
+            newContext = contextCache.getContext(newValue.schemaId);
+        }
+
+        // Use row encoder for aggregation
+        RowEncoder encoder = newContext.getRowEncoder();
+        encoder.startNewRow();
+        AggregateFieldsProcessor.aggregateAllFields(
+                oldValue.row, newValue.row, oldContext, newContext, encoder);
+        BinaryRow mergedRow = encoder.finishRow();
+
+        return new BinaryValue(newValue.schemaId, mergedRow);
```

Comment:
Why not use `targetSchemaId` `RowEncoder` to encode the new row? The `newValue.schemaId` is input data which is defined by client, and is usually outdated.